### PR TITLE
Kati-21 [Server] Implement Facilities/Specialties/DegreeType for Database

### PR DESCRIPTION
**Pull request by huuduy216**

### CODE REVIEW

File: Servers/.settings/org.eclipse.wst.server.core.prefs [hunk 1]
```diff
@@ -1,3 +1,3 @@
-#Thu May 26 13:04:04 ICT 2016
```
Comment (hieumai) [inline Servers/.settings/org.eclipse.wst.server.core.prefs]: Ignore this file

File: Servers/Tomcat v7.0 Server at localhost-config/server.xml [hunk 2]
```diff
@@ -65,7 +65,7 @@
     <!-- A "Connector" using the shared thread pool-->
     <!--
     <Connector executor="tomcatThreadPool"
-               port="8080" protocol="HTTP/1.1"
+               port="8085" protocol="HTTP/1.1"
```
Comment (hieumai) [inline Servers/Tomcat v7.0 Server at localhost-config/server.xml]: Maybe we should ignore the whole Server project

Comment (huuduy216) [inline Servers/Tomcat v7.0 Server at localhost-config/server.xml]: We could, I've been investigating the folder and it seems like only Eclipse uses it to run the server. Therefore there should be no impact on the project.

File: com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	
+	public ResultSet queryAll(String tableName) throws SQLException {
+		PreparedStatement stmt =null;
+		ResultSet rs = null;
+		stmt = connectToDatabase();
+		rs = stmt.executeQuery("SELECT * FROM " + tableName);
+		return rs;
+	}
+
+	public ResultSet queryFacilityOpeningHours(String id) throws SQLException {
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java]: First of all, the method name seems wrong (maybe queryOpeningHoursByFacilityID )
Second of all, putting Facility in DatabaseHelper is bad ( too specific )

File: com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+			stmt = con.prepareStatement("SELECT * FROM opening_hours WHERE facility_id = ?");
+			stmt.setInt(1,Integer.parseInt(id));
+			rs = stmt.executeQuery();
+			return rs;
+		} finally {
+
+		}
+	}
+
+	public ResultSet queryFacilityByName(String name) throws SQLException {
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java]: queryFacilityByName should not be in Database Helper too

File: com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+		try {
+
+			stmt = connectToDatabase();
+			rs = stmt.executeQuery("SELECT * FROM Facilities WHERE name = \"" + name + "\"");
+			return rs;
+		} finally {
+
+		}
+	}
+	public ResultSet queryFacilityByID(String id) throws SQLException {
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java]: queryFacilityByName should not be in Database Helper too

File: com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+		}
+	}
+
+	public ResultSet queryFacilityByName(String name) throws SQLException {
+		PreparedStatement stmt =null;
+		ResultSet rs = null;
+		try {
+
+			stmt = connectToDatabase();
+			rs = stmt.executeQuery("SELECT * FROM Facilities WHERE name = \"" + name + "\"");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java]: use prepareStatement and setString for this

File: com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+		}
+	}
+	public ResultSet queryFacilityByID(String id) throws SQLException {
+		PreparedStatement stmt =null;
+		ResultSet rs = null;
+		try {
+
+			stmt = connectToDatabase();
+			rs = stmt.executeQuery("SELECT * FROM Facilities WHERE facility_id = \"" + id + "\"");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java]: use prepareStatement and setString for this

File: com.kms.cura.dal/src/com/kms/cura/dal/DegreeDAL.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+		return _instance;
+	}
+
+	public List<Entity> getAll() throws ClassNotFoundException, SQLException {
+		return super.getAll(DEGREE_TABLE_NAME);
+	}
+
+	@Override
+	protected Entity getEntityFromResultSet(ResultSet resultSet, DatabaseHelper dbh) throws SQLException {
+		// TODO Auto-generated method stub
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/DegreeDAL.java]: remove this comment

File: com.kms.cura.dal/src/com/kms/cura/dal/DegreeDAL.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+		return new DegreeEntity(resultSet.getString("degree_id"),resultSet.getString("name"));
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/DegreeDAL.java]: check if result set is null first before getting data

File: com.kms.cura.dal/src/com/kms/cura/dal/EntityDAL.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+			}
+			return result;
+		} finally {
+			if (dbh != null) {
+				dbh.closeConnection();
+			}
+		}
+	}
+
+	public static FacilityEntity getFacilityByName(String name) throws SQLException, ClassNotFoundException {
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/EntityDAL.java]: should be in FacilityDAL

File: com.kms.cura.dal/src/com/kms/cura/dal/EntityDAL.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+			}
+			return result;
+		} finally {
+			if (dbh != null) {
+				dbh.closeConnection();
+			}
+		}
+	}
+
+	public static FacilityEntity getFacilityByID(String id) throws SQLException, ClassNotFoundException {
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/EntityDAL.java]: should be in FacilityDAL

File: com.kms.cura.dal/src/com/kms/cura/dal/EntityDAL.java (hunk 10 above)
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/EntityDAL.java]: Also, write a getEntityByName method and abstract this

File: com.kms.cura.dal/src/com/kms/cura/dal/EntityDAL.java (hunk 11 above)
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/EntityDAL.java]: Also, write a getEntityByID method and abstract this

File: com.kms.cura.dal/src/com/kms/cura/dal/SpecialityDAL.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+		return _instance;
+	}
+
+	public List<Entity> getAll() throws ClassNotFoundException, SQLException {
+		return super.getAll(SPECIALITY_TABLE_NAME);
+	}
+
+	@Override
+	protected Entity getEntityFromResultSet(ResultSet resultSet, DatabaseHelper dbh) throws SQLException {
+		// TODO Auto-generated method stub
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/SpecialityDAL.java]: Remove generated comment

File: com.kms.cura.dal/src/com/kms/cura/dal/SpecialityDAL.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+		return new SpecialityEntity(resultSet.getString("speciality_id"),resultSet.getString("name"));
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/SpecialityDAL.java]: check resultSet first before getting data

File: com.kms.cura.dal/target/classes/META-INF/maven/com.kms.cura.dal/com.kms.cura.dal/pom.properties [hunk 14]
```diff
@@ -0,0 +1,7 @@
+#Generated by Maven Integration for Eclipse
```
Comment (hieumai) [inline com.kms.cura.dal/target/classes/META-INF/maven/com.kms.cura.dal/com.kms.cura.dal/pom.properties]: Ignore this file

File: com.kms.cura.entity/src/com/kms/cura/entity/user/AdminUserEntity.java [hunk 15]
```diff
@@ -0,0 +1,9 @@
+package com.kms.cura.entity.user;
+
+public class AdminUserEntity extends UserEntity{
```
Comment (hieumai) [inline com.kms.cura.entity/src/com/kms/cura/entity/user/AdminUserEntity.java]: Re-format the file

File: com.kms.cura.entity/src/com/kms/cura/entity/user/DoctorUserEntity.java [hunk 16]
```diff
@@ -0,0 +1,106 @@
+package com.kms.cura.entity.user;
+
+import com.kms.cura.entity.DegreeEntity;
+import com.kms.cura.entity.FacilityEntity;
+import com.kms.cura.entity.SpecialityEntity;
+import java.sql.Date;
+
+
+public class DoctorUserEntity extends UserEntity{
```
Comment (hieumai) [inline com.kms.cura.entity/src/com/kms/cura/entity/user/DoctorUserEntity.java]: Re-format the file

File: com.kms.cura.entity/src/com/kms/cura/entity/user/DoctorUserEntity.java [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+	private String phone;
+	private DegreeEntity degree;
+	private SpecialityEntity speciality;
+	private double rating;
+	private int experience;
+	private double minPrice;
+	private double maxPrice;
+	private FacilityEntity facility_id;
```
Comment (hieumai) [inline com.kms.cura.entity/src/com/kms/cura/entity/user/DoctorUserEntity.java]: Should rename facility_id to facility, also doctor should be able to have multiple facilities

File: com.kms.cura.entity/src/com/kms/cura/entity/user/DoctorUserEntity.java [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+import com.kms.cura.entity.DegreeEntity;
+import com.kms.cura.entity.FacilityEntity;
+import com.kms.cura.entity.SpecialityEntity;
+import java.sql.Date;
+
+
+public class DoctorUserEntity extends UserEntity{
+	private String phone;
+	private DegreeEntity degree;
+	private SpecialityEntity speciality;
```
Comment (hieumai) [inline com.kms.cura.entity/src/com/kms/cura/entity/user/DoctorUserEntity.java]: Doctor should be able to have multiple specialities

File: com.kms.cura.entity/src/com/kms/cura/entity/user/PatientUserEntity.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
+package com.kms.cura.entity.user;
+
+import java.sql.Date;
+
+public class PatientUserEntity extends UserEntity{
```
Comment (hieumai) [inline com.kms.cura.entity/src/com/kms/cura/entity/user/PatientUserEntity.java]: re-format the file

File: com.kms.cura.entity/target/classes/META-INF/maven/com.kms.cura.entity/com.kms.cura.entity/pom.properties [hunk 20]
```diff
@@ -0,0 +1,7 @@
+#Generated by Maven Integration for Eclipse
```
Comment (hieumai) [inline com.kms.cura.entity/target/classes/META-INF/maven/com.kms.cura.entity/com.kms.cura.entity/pom.properties]: Ignore this file

File: com.kms.cura_server/.gitignore [hunk 21]
```diff
@@ -1,2 +1,3 @@
 /build/
 /target/
+/src/com/kms/cura_server/test.java
```
Comment (hieumai) [inline com.kms.cura_server/.gitignore]: Don't add this to gitignore. Keep it in your local and don't commit it instead

File: com.kms.cura_server/src/com/kms/cura_server/DegreeAPI.java [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+public class DegreeAPI {
+	@GET
+	@Path("/getAll")
+	public String getAllDegree() {
+		try {
+			List<Entity> degree = DegreeDAL.getInstance().getAll();
+			JsonElement element = EntityToJsonConverter.convertEntityListToJson(degree);
+			return element.toString();
+		} catch (ClassNotFoundException | SQLException e) {
+			// TODO Auto-generated catch block
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/DegreeAPI.java]: Remove this comment

File: com.kms.cura_server/src/com/kms/cura_server/DegreeAPI.java [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+			e.printStackTrace();
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/DegreeAPI.java]: Do not printstracktrace

File: com.kms.cura_server/src/com/kms/cura_server/FacilityAPI.java [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+public class FacilityAPI {
+	@GET
+	@Path("/getAll")
+	public String getAllFacility() {
+		try {
+			List<Entity> facilities = FacilityDAL.getInstance().getAll();
+			JsonElement element = EntityToJsonConverter.convertEntityListToJson(facilities);
+			return element.toString();
+		} catch (ClassNotFoundException | SQLException e) {
+			// TODO Auto-generated catch block
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/FacilityAPI.java]: Remove this comment

File: com.kms.cura_server/src/com/kms/cura_server/FacilityAPI.java [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+			e.printStackTrace();
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/FacilityAPI.java]: Don't printStackTrace

File: com.kms.cura_server/src/com/kms/cura_server/SpecialityAPI.java [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+public class SpecialityAPI {
+	@GET
+	@Path("/getAll")
+	public String getAllDegree() {
+		try {
+			List<Entity> speciality = SpecialityDAL.getInstance().getAll();
+			JsonElement element = EntityToJsonConverter.convertEntityListToJson(speciality);
+			return element.toString();
+		} catch (ClassNotFoundException | SQLException e) {
+			// TODO Auto-generated catch block
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/SpecialityAPI.java]: Remove this comment

File: com.kms.cura_server/src/com/kms/cura_server/SpecialityAPI.java [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+			e.printStackTrace();
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/SpecialityAPI.java]: Remote printStackTrace

File: Servers/Tomcat v7.0 Server at localhost-config/server.xml (hunk 2 above)
Comment (hieumai) [inline Servers/Tomcat v7.0 Server at localhost-config/server.xml]: So you should remove the Servers project and add it to gitignore

File: com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+		ResultSet rs = null;
+		stmt = connectToDatabase();
+		stmt = con.prepareStatement("SELECT * FROM ? WHERE name = ?");
+		stmt.setString(1, tableName);
+		stmt.setString(2, name);
+		rs = stmt.executeQuery();
+		return rs;
+	}
+
+	public ResultSet queryOpeningHoursByFacilityID(String id) throws SQLException {
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/DatabaseHelper.java]: queryOpeningHoursByFacilityID is too specific for DatabaseHelper. You can abstract it to queryByReferenceId(String tableName, String referenceIdKey, String referenceIdValue)

File: com.kms.cura.dal/src/com/kms/cura/dal/InterfaceAdapter.java [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+
+public class InterfaceAdapter<T>
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/InterfaceAdapter.java]: What is this class used for ?

File: com.kms.cura_server/src/com/kms/cura_server/SpecialityAPI.java [hunk 30]
```diff
@@ -0,0 +1,28 @@
...
+public class SpecialityAPI {
+	@GET
+	@Path("/getAll")
+	public String getAllDegree() {
+		try {
+			List<Entity> speciality = SpecialityDAL.getInstance().getAll();
+			JsonElement element = EntityToJsonConverter.convertEntityListToJson(speciality);
+			return element.toString();
+		} catch (ClassNotFoundException | SQLException e) {
+			return "Internal Error";
```
Comment (hieumai) [inline com.kms.cura_server/src/com/kms/cura_server/SpecialityAPI.java]: Extract String to maybe a class to reuse later.

**Pull request merged**